We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that the number of users of our apps determines our revenue for any given app — the more users who see and engage with the ads, the better. Our goal for this project is to analyze data to help our developers understand what type of apps are likely to attract more users.

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
from csv import reader

# The Google Play data set
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

# The App Store data set
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

In [4]:
print(android_header)
print('\n')
explore_data(android,1,3)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']




In [5]:
print(ios_header)
print('\n')
explore_data(ios,1,3,rows_and_columns=True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


### missing data in row 10472

In [6]:
android[10472]

['Life Made WI-Fi Touchscreen Photo Frame',
 '1.9',
 '19',
 '3.0M',
 '1,000+',
 'Free',
 '0',
 'Everyone',
 '',
 'February 11, 2018',
 '1.0.19',
 '4.0 and up']

In [7]:
del android[10472]

### searching for duplicate apps in IOS dataset

In [8]:
ios_unique_apps = [] 
ios_duplicate_apps = [] 

for app in ios: 
    app_name = app[1] 

    if app_name not in ios_unique_apps:
        ios_unique_apps.append(app_name)
    else:
        ios_duplicate_apps.append(app)

In [33]:
#ios_duplicate_apps

In [32]:
len(ios_unique_apps)

7195

In [10]:
for apps in ios:
    if apps[1]=='Mannequin Challenge':
        print(apps)
print('\n')
for apps in ios:
    if apps[1]=='VR Roller Coaster':
        print(apps)

['1173990889', 'Mannequin Challenge', '109705216', 'USD', '0.0', '668', '87', '3.0', '3.0', '1.4', '9+', 'Games', '37', '4', '1', '1']
['1178454060', 'Mannequin Challenge', '59572224', 'USD', '0.0', '105', '58', '4.0', '4.5', '1.0.1', '4+', 'Games', '38', '5', '1', '1']


['952877179', 'VR Roller Coaster', '169523200', 'USD', '0.0', '107', '102', '3.5', '3.5', '2.0.0', '4+', 'Games', '37', '5', '1', '1']
['1089824278', 'VR Roller Coaster', '240964608', 'USD', '0.0', '67', '44', '3.5', '4.0', '0.81', '4+', 'Games', '38', '0', '1', '1']


Even though they have the same names, both app names two apps each, but i do not think they are the same apps as they have different donwload size

### duplicate entries for android

In [11]:
android_unique_apps = [] 
android_duplicate_apps = [] 

for app in android: 
    app_name = app[0] 

    if app_name not in android_unique_apps:
        android_unique_apps.append(app_name)
    else:
        android_duplicate_apps.append(app_name)

In [47]:
#android_duplicate_apps

In [39]:
len(android_duplicate_apps)

1181

In [48]:
for apps in android:
    if apps[0]=='Instagram':
        print(apps)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


There's 1181 entries that are duplicates, might need to delete duplicate entries by selecting ones taht are most recent by choosing those with highest amount of ratings

creating a dictionary with unique app names with the highest ratings (if there were duplicates)

In [12]:
reviews_max={}
for row in android:
    name=row[0]
    n_reviews=float(row[3])
    if name in reviews_max and reviews_max[name]<n_reviews:
        reviews_max[name]=n_reviews
    elif name not in reviews_max:
        reviews_max[name]=n_reviews

In [60]:
len(reviews_max)

9659

Removing duplicate rows

In [13]:
android_clean=[]
already_added=[]

for row in android:
    name=row[0]
    n_reviews=float(row[3])
    if n_reviews==reviews_max[name] and (name not in already_added):
        android_clean.append(row)
        already_added.append(name)

In [19]:
android.header()

AttributeError: 'list' object has no attribute 'header'

In [14]:
def check_english(string):
    for letter in string:
        #ascii characters >127 means it's not an english letter/symbol
        if ord(letter)>127 :
            return False
        return True

In [15]:
print(check_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(check_english('Docs To Go™ Free Office Suite'))
print(check_english('Instachat 😜'))

False
True
True


this filter function might not be perfect cause having emojis which have will have an ascii value of more than 127 will make it not english using the function(check_english)
Need to create new one which limits amount of n

In [16]:
def check_english_v2(string):
    non_eng_count = 0
    for letter in string:
        #ascii characters >127 - it's not an english letter/symbol
        if ord(letter)>127 :
            non_eng_count +=1
        
    if non_eng_count >3:
        return False
    else: 
        return True   

In [17]:
print(check_english_v2('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(check_english_v2('Docs To Go™ Free Office Suite'))
print(check_english_v2('Instachat 😜'))

False
True
True


In [20]:
android_clean_english = []
for row in android_clean:
    name = row[0]
    if check_english_v2(name):
        android_clean_english.append(row)
        
    
ios_english = []
for row in ios:
    name = row[1]
    if check_english_v2(name):
        ios_english.append(row)

Now we want to isolate the free apps

In [24]:
android[2][7]

'0'

In [25]:
free_android_clean_english = []
for row in android_clean_english:
    price = row[7]
    if price == '0':
        free_android_clean_english.append(row)

free_ios_english = []
for row in ios_english:
    price = row[4]
    if price == '0.0':
        free_ios_english.append(row)

In [26]:
len(free_android_clean_english)

8864

In [27]:
len(free_ios_english)

3222

There's only 8864 apps for android and 3222 apps for ios after we clean the data

To make it easier for future code, i'll change the variables of final versions of each dataset to a shorter one

In [29]:
android_final = free_android_clean_english
ios_final = free_ios_english

As we mentioned in the introduction, our goal is to determine the kinds of apps that are likely to attract more users because the number of people using our apps affect our revenue.

To minimize risks and overhead, our validation strategy for an app idea has three steps:

Build a minimal Android version of the app, and add it to Google Play.
If the app has a good response from users, we develop it further.
If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.
Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful in both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by determining the most common genres for each market. For this, we'll need to build frequency tables for a few columns in our datasets.

### Building a frequency table to identify which genres & categories are most popular in each dataset

In [32]:
#build a percentages table
def frequency_table(dataset,index):
    table={}
    
    for row in dataset:
        if row[index] in table:
            table[row[index]] += 1
        else:
            table[row[index]] = 1
            
    percentages_table = {}
    for key in table:
        percentage = ((table[key])/len(dataset))*100
        percentages_table[key]=percentage
        
    return percentages_table

In [39]:
#create function to sort dictionary in descending order based on its keys
def sort_dict_desc(dataset,index):
    freq = frequency_table(dataset,index)
    sort_orders = sorted(freq.items(), key=lambda x: x[1], reverse=True)

    for i in sort_orders:
        print(i[0], i[1])

In [42]:
#ios prime_genre 
sort_dict_desc(ios_final,-5)

Games 58.16263190564867
Entertainment 7.883302296710118
Photo & Video 4.9658597144630665
Education 3.662321539416512
Social Networking 3.2898820608317814
Shopping 2.60707635009311
Utilities 2.5139664804469275
Sports 2.1415270018621975
Music 2.0484171322160147
Health & Fitness 2.0173805090006205
Productivity 1.7380509000620732
Lifestyle 1.5828677839851024
News 1.3345747982619491
Travel 1.2414649286157666
Finance 1.1173184357541899
Weather 0.8690254500310366
Food & Drink 0.8069522036002483
Reference 0.5586592178770949
Business 0.5276225946617008
Book 0.4345127250155183
Navigation 0.186219739292365
Medical 0.186219739292365
Catalogs 0.12414649286157665


We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

In [45]:
# android Category
sort_dict_desc(android_final, 1)

FAMILY 18.907942238267147
GAME 9.724729241877256
TOOLS 8.461191335740072
BUSINESS 4.591606498194946
LIFESTYLE 3.9034296028880866
PRODUCTIVITY 3.892148014440433
FINANCE 3.7003610108303246
MEDICAL 3.531137184115524
SPORTS 3.395758122743682
PERSONALIZATION 3.3167870036101084
COMMUNICATION 3.2378158844765346
HEALTH_AND_FITNESS 3.0798736462093865
PHOTOGRAPHY 2.944494584837545
NEWS_AND_MAGAZINES 2.7978339350180503
SOCIAL 2.6624548736462095
TRAVEL_AND_LOCAL 2.33528880866426
SHOPPING 2.2450361010830324
BOOKS_AND_REFERENCE 2.1435018050541514
DATING 1.861462093862816
VIDEO_PLAYERS 1.7937725631768955
MAPS_AND_NAVIGATION 1.3989169675090252
FOOD_AND_DRINK 1.2409747292418771
EDUCATION 1.1620036101083033
ENTERTAINMENT 0.9589350180505415
LIBRARIES_AND_DEMO 0.9363718411552346
AUTO_AND_VEHICLES 0.9250902527075812
HOUSE_AND_HOME 0.8235559566787004
WEATHER 0.8009927797833934
EVENTS 0.7107400722021661
PARENTING 0.6543321299638989
ART_AND_DESIGN 0.6430505415162455
COMICS 0.6204873646209386
BEAUTY 0.59792418

Further investogation shows that Family Genre is mostly filled by game apps, so the games category also dominates the google play store dataset

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

In [46]:
#prime_genre = -5
#rating_count_tot = 5
iosgenres = frequency_table(ios_final,-5)

for genre in iosgenres:
    total=0
    len_genre=0
    for app in ios_final:
        genre_app=app[-5]
        if genre_app==genre:
            n_user_ratings = float(app[5])
            total+=n_user_ratings
            len_genre+=1
    
    avg_ratings = total/len_genre
    print(genre, ': ', avg_ratings)

Social Networking :  71548.34905660378
Photo & Video :  28441.54375
Games :  22788.6696905016
Music :  57326.530303030304
Reference :  74942.11111111111
Health & Fitness :  23298.015384615384
Weather :  52279.892857142855
Utilities :  18684.456790123455
Travel :  28243.8
Shopping :  26919.690476190477
News :  21248.023255813954
Navigation :  86090.33333333333
Lifestyle :  16485.764705882353
Entertainment :  14029.830708661417
Food & Drink :  33333.92307692308
Sports :  23008.898550724636
Book :  39758.5
Finance :  31467.944444444445
Education :  7003.983050847458
Productivity :  21028.410714285714
Business :  7491.117647058823
Catalogs :  4004.0
Medical :  612.0


Navigation Apps have highest avg ratings, might be influenced by Waze and Google Maps


In [49]:
for app in ios_final:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Might be easier to break into categories with less dominance by famous apps such as the productivity category

In [51]:
#showing productivity apps distribution
for app in ios_final:
    if app[-5] == 'Productivity':
        print(app[1], ':', app[5])

Evernote - stay organized : 161065
Gmail - email by Google: secure, fast & organized : 135962
iTranslate - Language Translator & Dictionary : 123215
Yahoo Mail - Keeps You Organized! : 113709
Google Docs : 64259
Google Drive - free online storage : 59255
Dropbox : 49578
Microsoft Word : 47999
Microsoft OneNote : 39638
Microsoft Outlook - email and calendar : 32807
Hotspot Shield Free VPN Proxy & Wi-Fi Privacy : 32499
Documents 6 - File manager, PDF reader and browser : 29110
Google Sheets : 24602
Microsoft Excel : 24430
Inbox by Gmail : 21561
T-Mobile : 19977
Paper by FiftyThree - Sketch, Diagram, Take Notes : 18219
MyScript Calculator - Handwriting calculator : 16555
VPN Proxy Master - Unlimited WiFi security VPN : 13674
Microsoft OneDrive – File & photo cloud storage : 12797
Ever - Capture Your Memories : 12755
Speak & Translate － Voice and Text Translator : 12062
Tayasui Sketches : 11505
Drawing Desk - Draw, Paint, Doodle & Sketch board : 11040
Microsoft PowerPoint : 10939
Email - F

On the previous screen, we came up with an app profile recommendation for the App Store based on the number of user ratings. We have data about the number of installs for the Google Play market, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.)

In [47]:
frequency_table(android_final,5)

{'10,000+': 10.198555956678701,
 '5,000,000+': 6.825361010830325,
 '50,000,000+': 2.3014440433213,
 '100,000+': 11.552346570397113,
 '50,000+': 4.7721119133574,
 '1,000,000+': 15.726534296028879,
 '10,000,000+': 10.548285198555957,
 '5,000+': 4.512635379061372,
 '500,000+': 5.561823104693141,
 '1,000,000,000+': 0.22563176895306858,
 '100,000,000+': 2.1322202166064983,
 '1,000+': 8.393501805054152,
 '500,000,000+': 0.2707581227436823,
 '500+': 3.2490974729241873,
 '100+': 6.915613718411552,
 '50+': 1.917870036101083,
 '10+': 3.5424187725631766,
 '1+': 0.5076714801444043,
 '5+': 0.78971119133574,
 '0+': 0.04512635379061372,
 '0': 0.01128158844765343}

For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to find out which app genres attract the most users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on. To perform computations, however, we'll need to convert each install number from a string to a float. This means we need to remove the commas and the plus characters, or the conversion will fail and cause an error.

In [53]:
sort_dict_desc(android_final,5)

1,000,000+ 15.726534296028879
100,000+ 11.552346570397113
10,000,000+ 10.548285198555957
10,000+ 10.198555956678701
1,000+ 8.393501805054152
100+ 6.915613718411552
5,000,000+ 6.825361010830325
500,000+ 5.561823104693141
50,000+ 4.7721119133574
5,000+ 4.512635379061372
10+ 3.5424187725631766
500+ 3.2490974729241873
50,000,000+ 2.3014440433213
100,000,000+ 2.1322202166064983
50+ 1.917870036101083
5+ 0.78971119133574
1+ 0.5076714801444043
500,000,000+ 0.2707581227436823
1,000,000,000+ 0.22563176895306858
0+ 0.04512635379061372
0 0.01128158844765343


In [54]:
categories_android = frequency_table(android_final, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:



In [55]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])


WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:

In [57]:
under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3603485.3884615386

It's the same for most of the other categories cause genres like social media even though has high average ratings, still has some dominant apps that make teh data skewed like Instagram, Facebook, etc.


Books and Reference category might be 